src/kudu/consensus/leader_election.cc

```diff
@@ -138,6 +138,113 @@ bool VoteCounter::AreAllVotesIn() const {
   return GetTotalVotesCounted() == num_voters_;
 }
 
+void FlexibleVoteCounter::FetchTopologyInfo(
+    std::map<std::string, int>& voter_distribution,
+    std::map<std::string, int>& data_quorum_requirement) {
+  voter_distribution.clear();
+  data_quorum_requirement.clear();
+  // TODO(ritwikyadav): Populate the out params by reading config.
+}
+
+FlexibleVoteCounter::FlexibleVoteCounter(RaftConfigPB config)
+  : VoteCounter(0, 0),
+    config_(std::move(config)) {
+  num_voters_ = 0;
+  yes_vote_count_.clear();
+  no_vote_count_.clear();
+  FetchTopologyInfo(voter_distribution_, data_quorum_requirement_);
+  for (auto regional_vote_count : voter_distribution_) {
+    CHECK_GT(regional_vote_count.second, 0);
+    num_voters_ += regional_vote_count.second;
+    yes_vote_count_.emplace(regional_vote_count.first, 0);
+    no_vote_count_.emplace(regional_vote_count.first, 0);
+  }
+
+  CHECK_GT(num_voters_, 0);
+
+  bool non_empty_quorum = false;
+  for (auto regional_quorum_count : data_quorum_requirement_) {
+    auto total_voters = voter_distribution_.find(regional_quorum_count.first);
+    CHECK(total_voters != voter_distribution_.end());
+    CHECK_LE(regional_quorum_count.second, total_voters->second);
+    if (regional_quorum_count.second > 0) {
+      non_empty_quorum = true;
+    }
+  }
+  CHECK(non_empty_quorum);
```

Comment:
these validators need to become an API which should be called from Plugin.
No point crashing the server for wrong data, but validate it on entry.

Reply:
I agree, this should probably be moved into a separate validate() method.